internal/controller/k0smotron.io/k0smotroncluster_service.go

```diff
@@ -80,51 +94,60 @@ func (r *ClusterReconciler) reconcileServices(ctx context.Context, kmc km.Cluste
 	// Depending on ingress configuration create nodePort service.
 	logger.Info("Reconciling services")
 	svc := r.generateService(&kmc)
-	if kmc.Spec.Service.Type == v1.ServiceTypeLoadBalancer && kmc.Spec.ExternalAddress == "" {
-		err := r.watchLBServiceIP(ctx, kmc)
-		if err != nil {
-			return err
-		}
-	}
-	return r.Client.Patch(ctx, &svc, client.Apply, patchOpts...)
-}
-
-func (r *ClusterReconciler) watchLBServiceIP(ctx context.Context, kmc km.Cluster) error {
-	var handler cache.ResourceEventHandlerFuncs
-	stopCh := make(chan struct{})
-	handler.UpdateFunc = func(old, new interface{}) {
-		newObj, ok := new.(*v1.Service)
-		if !ok {
-			return
-		}
 
-		var externalAddress string
-		if len(newObj.Status.LoadBalancer.Ingress) > 0 {
-			if newObj.Status.LoadBalancer.Ingress[0].Hostname != "" {
-				externalAddress = newObj.Status.LoadBalancer.Ingress[0].Hostname
-			}
-			if newObj.Status.LoadBalancer.Ingress[0].IP != "" {
-				externalAddress = newObj.Status.LoadBalancer.Ingress[0].IP
+	if err := r.Client.Patch(ctx, &svc, client.Apply, patchOpts...); err != nil {
+		return err
+	}
+	// Wait for LB address to be available
+	logger.Info("Waiting for loadbalancer address")
+	if kmc.Spec.Service.Type == v1.ServiceTypeLoadBalancer && kmc.Spec.ExternalAddress == "" {
+		err := wait.PollUntilContextTimeout(ctx, 1*time.Second, 2*time.Minute, true, func(ctx context.Context) (bool, error) {
+			err := r.Client.Get(ctx, client.ObjectKey{Name: svc.Name, Namespace: svc.Namespace}, &svc)
+			if err != nil {
+				return false, err
 			}
+			if len(svc.Status.LoadBalancer.Ingress) > 0 {
+				if svc.Status.LoadBalancer.Ingress[0].Hostname != "" {
+					kmc.Spec.ExternalAddress = svc.Status.LoadBalancer.Ingress[0].Hostname
+				}
+				if svc.Status.LoadBalancer.Ingress[0].IP != "" {
+					kmc.Spec.ExternalAddress = svc.Status.LoadBalancer.Ingress[0].IP
+				}
+				logger.Info("Loadbalancer address available, updating Cluster object", "address", kmc.Spec.ExternalAddress)
 
-			if externalAddress != "" {
-				kmc.Spec.ExternalAddress = externalAddress
 				err := r.Client.Update(ctx, &kmc)
 				if err != nil {
-					log.FromContext(ctx).Error(err, "failed to update K0smotronCluster")
+					return false, err
 				}
-				stopCh <- struct{}{}
+				return true, nil
+			}
+			return false, nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get loadbalancer address: %w", err)
+		}
+	} else if kmc.Spec.Service.Type == v1.ServiceTypeNodePort && kmc.Spec.ExternalAddress == "" {
+		// Get a random node address as external address
+		nodes, err := r.ClientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		if len(nodes.Items) == 0 {
+			return fmt.Errorf("no nodes found")
+		}
+		// Get random node from list
+		node := nodes.Items[rand.Intn(len(nodes.Items))]
+		for _, addr := range node.Status.Addresses {
```

Comment:
Here we rely on the order of the addresses. Is it correct?

Reply:
Well, sort of. It _should_ prefer `NodeExternalIP` but I see what you mean. In some cases if it sees the internal one on the list first it'll pick it up. I can work it bit more to pick up both types of addresses and return one preferring external. I'll move this to a separate function too.